### PR TITLE
Feat/resend stored anchor requests on node startup CDB-2010

### PR DESCRIPTION
**Pull request by pawartur**
## Description

* Implemented the new `AnchorResumingService` class  with a single async `resumeRunningStatesFromAnchorRequestStore(...)` method which:
  * fetches stored anchor requests in batches
  * goes through each stored anchor request from each batch and adds a job to a concurrency queue, which
    * loads the stream related to the stored anchor request (which make sure that there's a running state for it)
    * re-requests the anchor for the same tip that was stored in the stored anchor
* started calling the new `resumeRunningStatesFromAnchorRequestStore(...)` method in a non-blocking way in `Ceramic._init(...)` method
* made sure that the concurrency queue owned by the new method is cleared when ceramic core is closed
* added tests that check that:
  * the new method is called by `Ceramic._init(...)` and is non-blocking
  * the new method re-requests anchors for stored anchor requests
* Started using the anchor request store to determine, if repository should start polling for anchor status updates when loading a running stream from the stream state store

## How Has This Been Tested?

Describe the tests that you ran to verify your changes. Provide instructions for reproduction.

- [X] Ran the automatic tests
- [X] Tested manually against S3

### CODE REVIEW

**Comment by pawartur** [comment]
TODOs:
1. Don't make the resume method blocking on daemon startup
2. Handle kill signal gracefully (see https://github.com/ceramicnetwork/js-ceramic/blob/57a25fa37ccdac3d378de847b7ce2f83200f390b/packages/cli/src/daemon/handle-sigint-signal.ts)

File: packages/core/src/state-management/repository.ts
Comment (ukstv) [inline packages/core/src/state-management/repository.ts:304]: Nice!

File: packages/core/src/state-management/repository.ts
Comment (ukstv) [inline packages/core/src/state-management/repository.ts:296]: I am a bit conflicted about this. Could you tell please how it relates to recovery mechanism in `fromStreamStateStore`?

File: packages/cli/src/ceramic-daemon.ts
Comment (ukstv) [inline packages/cli/src/ceramic-daemon.ts:306]: Since it is `async` operation, mind adding an error handler here, please?

Non-actionable: How about handling shutdown gracefully?

Comment (ukstv) [reply packages/cli/src/ceramic-daemon.ts:306]: UPD: Graceful shutdown is here. Have not seen that.

File: packages/core/src/state-management/repository.ts
Comment (ukstv) [inline packages/core/src/state-management/repository.ts:287]: Should it really be a part of repository?

Comment (ukstv) [reply packages/core/src/state-management/repository.ts:287]: IMO it would better be a separate service that uses the repository.

**Review by ukstv — CHANGES_REQUESTED** [review]
Would like to request some additional info, not really the changes.

1. Would like to have a short discussion about `resumeRunningStatesFromAnchorRequestStore` being part of repository. Maybe it could be own service. Repository is already too big.
2. Recovery in `fromStreamStateStore` and in `resumeRunningStatesFromAnchorRequestStore` - don't you think there is a duplication of requests to CAS?

File: packages/cli/src/s3-store.ts
Comment (pawartur) [reply packages/cli/src/s3-store.ts:92]: Ah, no, wait! We don't want to use `Partial` here and pass it to `store.createReadStream` because:
1. We don't want to pass the `useCaseName` to `store.createReadStream`
2. `StoreSearchParams` is a type defined specifically for the params passed to the search methods in ikv stores, so I think it's better to rely on the (non-)optionality of particular fields as defined in the type itself rather than make them all optional with `Partial`

Does that make sense, @ukstv ?

File: packages/core/src/state-management/repository.ts
Comment (pawartur) [reply packages/core/src/state-management/repository.ts:296]: I'm not sure how it relates to the recovery mechanism from the `fromStreamStateStore`, @ukstv . 

To me, this particular `if` block is just a sanity check, as there should be no anchor requests in storage for streams with anchor status `FAILED` or `ANCHORED` in the first place. I'm happy to remove this check here or replace it with a logger warning.

File: packages/core/src/state-management/repository.ts
Comment (pawartur) [reply packages/core/src/state-management/repository.ts:287]: Good call, will do!

**Comment by pawartur** [comment]
> Would like to request some additional info, not really the changes.
> 
> 1. Would like to have a short discussion about `resumeRunningStatesFromAnchorRequestStore` being part of repository. Maybe it could be own service. Repository is already too big.
> 2. Recovery in `fromStreamStateStore` and in `resumeRunningStatesFromAnchorRequestStore` - don't you think there is a duplication of requests to CAS?

Agree about 1.

When it comes to 2, we've agreed today with @ukstv that there should be no call to `anchor()` in the resuming method, but unfortunately we can't just remove it now, because we had found a bug that prevents streams to be properly updated in the stream state store after their anchor status is set to pending.

We will fix this bug in this branch and then remove the call to `anchor(...)` in the resuming method.

File: packages/cli/src/ceramic-daemon.ts
Comment (stbrody) [inline packages/cli/src/ceramic-daemon.ts:308]: I'm curious, is there a reason that the anchorResumingService is a part of the CeramicDaemon instead of the Ceramic core client?

File: packages/core/src/state-management/__tests__/anchor-resuming-service.test.ts
Comment (stbrody) [inline packages/core/src/state-management/__tests__/anchor-resuming-service.test.ts:75]: I'd love to not have to keep updating that PR for CDB-2090 as new PRs go in - any chance I could get a review on that soon and get it merged in so you could make the change here?

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (stbrody) [inline packages/core/src/state-management/anchor-resuming-service.ts:28]: is this line necessary?  It makes me think that shouldBeClosed could sometimes go from false to true back to false again.

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (stbrody) [inline packages/core/src/state-management/anchor-resuming-service.ts:45]: can you help me understand what this check is doing and why it's necessary?

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (stbrody) [inline packages/core/src/state-management/anchor-resuming-service.ts:47]: We need to be extremely careful here as if users delete their stream state store that could result in major data loss.  Is the intention actually to have them delete the anchor request store?  Even if we want them to just delete the anchor request store, we have to be very careful how we tell users to do this because if they get confused or do it wrong they could delete important data.

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (stbrody) [inline packages/core/src/state-management/anchor-resuming-service.ts:51]: Suggest using the verbose log level, as this could get quite noisy at startup otherwise if you have a large number of pending anchor requests.

File: packages/cli/src/ceramic-daemon.ts
Comment (pawartur) [reply packages/cli/src/ceramic-daemon.ts:308]: I did it this way, because every time I've been a part of discussing how this should work, we were saying that the resume should happen when the daemon is restarted.

Happy to move it to Ceramic core, if you think it should happen always when the core is client is started. I assume this should happen in `Ceramic._init(...)`, right?

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (pawartur) [reply packages/core/src/state-management/anchor-resuming-service.ts:28]: If the only way to get this running again after calling `close()` is to create a new instance, then it's not necessary. I'll remove it.

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (pawartur) [reply packages/core/src/state-management/anchor-resuming-service.ts:45]: When we were discussing this project during the spec review, you said that it was important for you to check, if the node's time is moving backwards. This is what this check is supposed to verify.

You think I should remove it or change it, @stbrody ?

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (pawartur) [reply packages/core/src/state-management/anchor-resuming-service.ts:47]: What would be a better message here, @stbrody (assuming that we want to keep this check)?

File: packages/cli/src/ceramic-daemon.ts
Comment (oed) [reply packages/cli/src/ceramic-daemon.ts:308]: That's right @pawartur. Agree that this makes more sense in core, because if you are using core directly this feature is still desirable.

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/memory/in-memory-anchor-service.ts:303]: wow, that previous behavior definitely seems wrong.  Kind of wild it never was an issue before now.  Good catch

File: packages/core/src/ceramic.ts
Comment (stbrody) [inline packages/core/src/ceramic.ts:551]:
```suggestion
      // We're not awaiting here on purpose, it's not supposed to be blocking
```

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (stbrody) [inline packages/core/src/state-management/anchor-resuming-service.ts:40]: nice catch changing from `repository.load()` to `repository.fromMemoryOrStore()`.  The former might have generated a lot of unnecessary pubsub traffic

File: packages/core/src/state-management/state-manager.ts
Comment (stbrody) [inline packages/core/src/state-management/state-manager.ts:194]: this will cause a lot of extra work to be done to re-pin every IPFS record in the entire stream log.  Is this to work around the issue where we don't update the state in the state store if the only thing that changed is the anchorStatus?  If so I think we need a more targeted fix for that.  Does that need to be fixed in this PR or could that be a follow-up?  Is it something we even need to worry about fixing right now?

File: packages/core/src/state-management/anchor-resuming-service.ts
Comment (stbrody) [reply packages/core/src/state-management/anchor-resuming-service.ts]: ah I see.  I guess I had imagined that being tracked and handled separately.  Tying the system clock check in with the data in the state store worries me for the reasons I mentioned in my other comment.  Can we have a separate ticket for the check around the time moving backward and do some brainstorming on the best and safest approach to address that?  That way we don't hold up this PR on trying to figure that all out

File: packages/core/src/indexing/__tests__/build-indexing.test.ts
Comment (stbrody) [inline packages/core/src/indexing/__tests__/build-indexing.test.ts:59]: probably should either just remove this TODO or update it with an actual ticket number before merging.

**Review by stbrody — APPROVED** [review]
LGTM

**Review by ukstv — APPROVED** [review]

**Pull request merged**
